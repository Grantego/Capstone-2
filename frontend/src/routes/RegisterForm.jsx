import React, {useState} from "react";
import { Form, redirect, useNavigate, useActionData } from "react-router-dom";
import SpellerApi from "../../api";
import {FormGroup, Label, Input, Button} from "reactstrap"
import NavBar from "../components/NavBar";
import "./body.css"

export async function action({request}) {
    try {
        const formData = await request.formData()
        const data = Object.fromEntries(formData)
        const errors = {}
        console.log(data)

        // validate the fields
        if (typeof data.username !== "string" || data.username.length < 4) {
        errors.username =
        "Username must be at least 4 characters";
        };

        if (typeof data.password !== "string" || data.password.length < 5) {
            errors.password = "Password must be at least 5 characters";
        };

        if (typeof data.email !== "string" || !data.email.includes("@")) {
            errors.email =
              "This is not an email address";
          }

        // return data if we have errors
        if (Object.keys(errors).length) {
            return errors;
        };
        const res = await SpellerApi.register(data)
        window.localStorage.setItem("username", data.username)
        window.localStorage.setItem("token", res.token)
        return redirect('/chords') 
    } catch(e) {
        alert(e)
        return redirect('/signup')
    }    
}

const RegisterForm = () => {

    let navigate = useNavigate()
    const errors = useActionData()
    return (
        <>
        <NavBar/>
        <div className="body-container">
            <h1>Register</h1>
            <Form method="post">
                <FormGroup>
                    <Label htmlFor="username">Username:</Label>
                    <Input 
                        type="text" 
                        name="username" 
                        id="username"
                        required
                        />
                    {errors?.username && <span>{errors.username}</span>}
                </FormGroup>
                <FormGroup>
                    <Label htmlFor="password">Password:</Label>
                    <Input 
                        type="password"
                        name="password"
                        id="password"
                        autoComplete="
                        new-password"
                        required/>
                        {errors?.password && <span>{errors.password}</span>}
                </FormGroup>
                <FormGroup>
                    <Label htmlFor="email">Email:</Label>
                    <Input 
                        type="text"
                        name="email" 
                        id="email"
                        required/>
                        {errors?.email && <span>{errors.email}</span>}
                </FormGroup>
                    <Button type="button" href="/">Cancel</Button>
                    <Button id="register-btn" type="submit">Register</Button>
            </Form>
            </div>
        </>
    )
}

export default RegisterForm