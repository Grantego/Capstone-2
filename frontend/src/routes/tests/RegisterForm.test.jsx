import React from 'react';
import { render, screen } from '@testing-library/react';
import RegisterForm from '../RegisterForm'
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import { describe, it, beforeEach, expect} from 'vitest';


describe('RegisterForm component', () => {
    let router

  beforeEach(() => {
    localStorage.clear();

    router = createMemoryRouter([
        {
            path: '/',
            element: <RegisterForm/>,
        },
    ], {
        initialEntries: ['/'],
    }); 


  });

  it("matches snapshot", () => {
    const {asFragment} = render(<RouterProvider router={router} />)
    expect(asFragment()).toMatchSnapshot();
  })

  it("renders without crashing", () => {
    render(<RouterProvider router={router} />)
  })

  it("renders correct elements", () => {
    render(<RouterProvider router={router} />)
    expect(screen.getByPlaceholderText("choose username")).toBeInTheDocument()    
    expect(screen.getByPlaceholderText("enter password")).toBeInTheDocument()    
    expect(screen.getByPlaceholderText("enter email")).toBeInTheDocument()    
    expect(screen.getByLabelText("Username:")).toBeInTheDocument()    
    expect(screen.getByLabelText("Password:")).toBeInTheDocument() 
    expect(screen.getByLabelText("Email:")).toBeInTheDocument()
    expect(screen.getAllByText("Register")).toBeInTheDocument
  })

});