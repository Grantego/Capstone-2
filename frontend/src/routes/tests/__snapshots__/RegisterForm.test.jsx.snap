// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChordDetails component > matches snapshot 1`] = `
<DocumentFragment>
  <nav
    class="navbar navbar-light bg-faded"
  >
    <div
      class="container-fluid"
    >
      <a
        class="me-auto navbar-brand"
        href="/"
      >
        Chord Helper
      </a>
      <button
        aria-label="Toggle navigation"
        class="me-2 navbar-toggler"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="collapse"
      >
        <ul
          class="navbar-nav"
        >
          <li
            class="nav-item"
          >
            <a
              class=""
              href="/signup"
            >
              Sign Up
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class=""
              href="/login"
            >
              Login
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
  <div
    class="body-container"
  >
    <h1>
      Register
    </h1>
    <form
      action="/"
      id="register-form"
      method="post"
    >
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="username"
        >
          Username:
        </label>
        <input
          class="form-control"
          id="username"
          name="username"
          placeholder="choose username"
          required=""
          type="text"
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="password"
        >
          Password:
        </label>
        <input
          autocomplete="
                        new-password"
          class="form-control"
          id="password"
          name="password"
          placeholder="enter password"
          required=""
          type="password"
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="email"
        >
          Email:
        </label>
        <input
          class="form-control"
          id="email"
          name="email"
          placeholder="enter email"
          required=""
          type="text"
        />
      </div>
      <a
        class="btn btn-secondary"
        href="/"
        type="button"
      >
        Cancel
      </a>
      <button
        class="btn btn-secondary"
        id="register-btn"
        type="submit"
      >
        Register
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`RegisterForm component > matches snapshot 1`] = `
<DocumentFragment>
  <nav
    class="navbar navbar-light bg-faded"
  >
    <div
      class="container-fluid"
    >
      <a
        class="me-auto navbar-brand"
        href="/"
      >
        Chord Helper
      </a>
      <button
        aria-label="Toggle navigation"
        class="me-2 navbar-toggler"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="collapse"
      >
        <ul
          class="navbar-nav"
        >
          <li
            class="nav-item"
          >
            <a
              class=""
              href="/signup"
            >
              Sign Up
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class=""
              href="/login"
            >
              Login
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
  <div
    class="body-container"
  >
    <h1>
      Register
    </h1>
    <form
      action="/"
      id="register-form"
      method="post"
    >
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="username"
        >
          Username:
        </label>
        <input
          class="form-control"
          id="username"
          name="username"
          placeholder="choose username"
          required=""
          type="text"
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="password"
        >
          Password:
        </label>
        <input
          autocomplete="
                        new-password"
          class="form-control"
          id="password"
          name="password"
          placeholder="enter password"
          required=""
          type="password"
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="email"
        >
          Email:
        </label>
        <input
          class="form-control"
          id="email"
          name="email"
          placeholder="enter email"
          required=""
          type="text"
        />
      </div>
      <a
        class="btn btn-secondary"
        href="/"
        type="button"
      >
        Cancel
      </a>
      <button
        class="btn btn-secondary"
        id="register-btn"
        type="submit"
      >
        Register
      </button>
    </form>
  </div>
</DocumentFragment>
`;
